gapii/client/capture.go

```diff
@@ -181,41 +186,109 @@ func (p *Process) Capture(ctx context.Context, start task.Signal, stop task.Sign
 	conn := p.conn
 	defer conn.Close()
 
-	var count siSize
-	started := false
+	if (p.Options.Flags & DeferStart) != 0 {
+		go func() {
+			if start.Wait(ctx) {
+				conn.Write([]byte{0x1, messageStartTrace})
+			}
+		}()
+	}
+
+	go func() {
+		if stop.Wait(ctx) {
+			conn.Write([]byte{0x1, messageEndTrace})
+		}
+	}()
+
+	buf := make([]byte, 64*1024)
+	var bufSize, bufPos int
+	var headerPos uint
+	var chunkSize uint64
+	var readErr error
+	var count, remain siSize
+	var messageType byte = messageInvalid
+	var bufErr []byte
+mainLoop:
 	for {
-		if task.Stopped(ctx) || stop.Fired() {
+		if task.Stopped(ctx) {
 			log.I(ctx, "Stop: %v", count)
-			break
+			break mainLoop
 		}
-		if (p.Options.Flags & DeferStart) != 0 {
-			if !started && start.Fired() {
-				started = true
-				w := endian.Writer(conn, device.LittleEndian)
-				w.Uint32(startMidExecutionCapture)
+		if bufPos >= bufSize {
+			// handle read error after processing buffer
+			switch {
+			case errors.Cause(readErr) == io.EOF:
+				// unexpected end of stream
+				log.E(ctx, "EOF: %v", count)
+				return int64(count), readErr
+			case readErr != nil && count > 0:
+				err, isnet := readErr.(net.Error)
+				if !isnet || (!err.Temporary() && !err.Timeout()) {
+					log.E(ctx, "Connection error: %v", err)
+					// Got an error mid-stream terminate.
+					return int64(count), err
+				}
+			case readErr != nil && count == 0:
+				// Got an error without receiving a byte of data.
+				// Treat failure-to-connect as target-not-ready instead of an error.
+				return 0, nil
 			}
-		}
-		now := time.Now()
-		conn.SetReadDeadline(now.Add(time.Millisecond * 500)) // Allow for stop event and UI refreshes.
-		n, err := io.CopyN(w, conn, 1024*64)
-		count += siSize(n)
-		atomic.StoreInt64(written, int64(count))
-		switch {
-		case errors.Cause(err) == io.EOF:
-			// End of stream. End.
-			log.I(ctx, "EOF: %v", count)
-			return int64(count), nil
-		case err != nil && count > 0:
-			err, isnet := err.(net.Error)
-			if !isnet || (!err.Temporary() && !err.Timeout()) {
-				log.I(ctx, "Connection error: %v", err)
-				// Got an error mid-stream terminate.
-				return int64(count), err
+			// read new buffer from connection
+			bufPos = 0
+			now := time.Now()
+			conn.SetReadDeadline(now.Add(time.Millisecond * 500)) // Allow for stop event and UI refreshes.
+			bufSize, readErr = conn.Read(buf)
+		} else if remain == 0 {
+			// read message header
+			if headerPos == 0 {
+				messageType = buf[bufPos]
+				if messageType == messageError {
+					bufErr = nil
+				}
+			} else {
+				chunkSize += uint64(buf[bufPos]) << ((headerPos - 1) * 8)
```

Comment:
These kind of lines looks cryptic to me, I'm thinking of someone else reading the code in a few months time: I would have a hard time understanding the "why" behind this line. Could it be split in a few more lines and commented a bit? e.g.:

```
// Header is 1 byte for type, then size in 8 bytes little-endian
if headerPos == 0 {
 // parse type
  [...]
} else {
  // parse size, one byte at a time
  byteIndex := (headerPos - 1)
  shiftAmount := byteIndex * 8
  sizeIncrement := uint64(buf[bufPos]) << shiftAmount
  chunkSize += sizeIncrement
}
```

Reply:
added a comment to explain that a deserialization is happening. This should be enough to explain the shifting.